Add Galaxy wrapper for BiaPy tool

danifranco opened the pull request:
Hi all,

I'm submitting this initial version of a Galaxy wrapper for [BiaPy](https://biapy.readthedocs.io/), a deep learning framework for bioimage analysis.
This tool is designed to run a complete BiaPy workflow with two main usage modes:

1. Use a custom YAML configuration: Users can upload their own YAML config file and run BiaPy directly.
2. Create the YAML interactively: The tool generates a config file based on user inputs through a set of guided questions (conditional logic).

I’ve been working on this for a few days and, as this is my first Galaxy tool, I have some questions and doubts that I would really appreciate your feedback on:

1. **Parameter help text**. Is there a better way to provide detailed help for parameters, other than using long help="..." strings directly in the XML? The inputs have quite a bit of context, so I'm wondering if there's a cleaner way to manage this.
2. **Conda packaging**. I’m in the process of adding BiaPy to `conda-forge` ([PR here](https://github.com/conda-forge/staged-recipes/pull/30597)). Once it’s accepted, can I declare it in the requirements section like this?

    ```
    <requirement type="package" version="3.6.2">biapy</requirement>
    ```

    using the `conda-forge` channel?
3. **GPU support**. BiaPy supports specifying a GPU index as a parameter (e.g., `--gpu 0`). Training without GPU is very slow, but inference is fine on CPU. What's the best practice in Galaxy for handling GPU support? Should I remove this parameter or keep it optional?
4. **Clickable links in parameter help**. Can we include URLs in `help=""` so they are rendered as clickable links in the Galaxy interface? I did not test how they look like, because there are still things to do in this PR so it passes the checks, and to try it locally, so maybe the are already displayed like links. 
5. **Conditional logic depending on two parameters**. I’d like to show or hide certain inputs based on the combination of two user choices. From [this issue](https://github.com/galaxyproject/galaxy/issues/13931), it seems that’s not yet supported. Is that still the case? If so, I’ve repeated some inputs unnecessarily to work around this, but it would be good to confirm.
6. **GUI-side "check consistency" button**. In our current GUI version of BiaPy (outside Galaxy), we allow users to click a button to validate dataset consistency after selecting input folders. Is there a Galaxy-native way to do this (e.g., a button, a validation script or pre-run check)?
7. **EDAM terms**. I'm not 100% sure the EDAM operations I selected are correct. Some aren’t even defined in EDAM, but are common DL tasks (like classification, image-to-image translation, etc.). Would it be better to refer to EDAM-bioimaging terms? Or just leave out undefined terms entirely?
8. **Creating the YAML config programmatically**. In the second usage mode, I need to generate the YAML file first, then pass it to the BiaPy command. Currently I use a Python script to generate it. Is that acceptable for Galaxy tool execution flow (i.e., intermediate file creation within the command section)?
9. **Testing output folders**. Since Galaxy doesn’t support a `type="directory"` output, I was advised (by ChatGPT) to compress the output folder into a .tar.gz file:

    ```
    <data name="results_archive" format="tar.gz" ... />
    ```
    Is this the correct way to handle directory outputs? 

    I had two test scenarios in mind:

    1. Inference using a custom config (`test_data/input_test_data/example.yaml`) and pretrained model (just to check basic output).
    2. Running with full interactive config creation, to verify YAML creation and complete execution.

    I wanted to validate the test by asserting the presence of specific files in the tar archive (e.g. checkpoints, predictions). The full output structure is in `test_data/my_experiment_example.tar.gz`, and the test data is in` test-data/input_test_data/`, but I wasn't sure how best to structure this in the test block. In general, I have doubts on how to configure the output and the tests cases so maybe you can guide me a little bit.

Thanks a lot!

@bgruening @iuc-member

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,316 @@
+        </section>
+        <section name="advanced_opts" title="Advanced options" expanded="False">
+            <param name="run_id" type="integer" value="1" min="1" max="100" optional="true" label="ID" help="Number that should be increased when one need to run the same job multiple times (reproducibility)"/>
+            <param name="gpu" type="text" optional="true" label="Select a GPU to use" help="Number of the GPU to run the job in"/>
```
A Galaxy user (typically) has no control on CPU/GPU/memory parameters this should be admin controllable.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,316 @@
+            </when>
+        </conditional>
+        <section name="general_opts" title="General options" expanded="True">
+            <param name="result_dir" type="text" label="Select the output directory" help="A directory path where all the workflow files will be stored. Just a folder name, not a full path"/>
```
This must not be an arbitrary string. Users could enter malicious text. Maybe use a regex validator. 

Does the user really need to control this?

danifranco replied:
In BiaPy the user typically sets a path to a folder to store the results of the workflow. If there is an automatic output folder in Galaxy maybe we can use that automatically? What happens if the users want to run a workflow more than once? the output folder is the same? this will overwrite all the files no?

bernt-matthias replied:
Every job runs in its own working dir and the tool can only write to this directory. So I guess `mkdir output` and hardcode `result_dir` as `output` is the solution.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,316 @@
+                            </section> 
+                            <section name="test_sec" title="Test data" expanded="True"> 
+                                <!-- Q11 -->
+                                <param name="test_raw_path" type="data" format="txt" label="Specify the location of the test raw image folder" help="In this step, you need to specify the folder where the images for testing the network are located. All these images must have the same number of channels. It's also important that each channel contains the same type of information to avoid confusing  ...
```
User should not have control over any folder names / paths. Instead use a data parameter where the images are given to the tool.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,316 @@
+                        <when value="train">
+                            <section name="train_sec" title="Train data" expanded="True"> 
+                                <!-- Q9 (repeated)--> 
+                                <param name="train_raw_path" type="data" format="txt" label="Specify the location of the training raw image folder" help="In this step, you need to specify the folder where the images for training the network are located. All these images must have the same number of channels. It's also important that each channel contains the same type of information to avoid conf ...
+                                <!-- Q10 (repeated)-->
+                                <param name="train_gt_path" type="data" format="txt" label="Specify the location of the training ground truth (target) folder" help="In this step, you need to specify the directory where the training target data, also known as 'ground truth', is located. This target data will be used to train the network. The target vary depending on the workflow: For Semantic segm ...
```
Do not provide paths, but images.

danifranco replied:
Answered in the above comment.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,316 @@
+                        </param>
+                        <when value="biapy_pretrained">
+                            <!-- Q4 -->
+                            <param name="biapy_model_path" type="data" format="data" label="Select the model trained with BiaPy before" help="This question is about setting the path to a pre-trained model trained before using BiaPy. The model should be stored in the folder you previously selected for saving results. Specifically, in the 'checkpoints' folder, there will be a file with a '.pth' ext ...
```
do we have a more specific format than data in Galaxy

danifranco replied:
Same here, I'm not sure how I need to handle this.

danifranco replied:
The model checkpoint expected here is in `state_dict` format (`.pth` extension). Do you have something for that or I can leave it as `data` as it is? I suppose that I would need to do the `ln -s` as it is done with images, no?

bernt-matthias replied:
Maybe `torch.onnx.export` .. seems that we already export `onnx`.

bernt-matthias replied:
> My question was more related to how to handle the file. If I need to proceed as with the training/test data.

Can you be more specific? 

> As there is no specific format for this file extension I can just put a generic data format as the one I was using I asumme. 

I need to stress that processing pth file (ie pickle files) is problematic. There is no problem if you work on private infrastructure and only have trustworthy users. But on public infrastructure it may become one because pickle files may allow arbitrary code execution. 

It would be really cool if biapy would support a safer format for saving the models.

danifranco replied:
> Can you be more specific?

In the same way you explained here: https://github.com/galaxyproject/tools-iuc/pull/7168#discussion_r2230773776 . Because this is the way Galaxy handles the input files, no?

Regarding the checkpoint we do not support any other extension but .pth. We load the weights as follows:
```
torch.load(checkpoint_path, weights_only=True, map_location="cpu")
```
This restricts deserialization to tensors only, blocking arbitrary Python object execution and greatly reducing pickle-related risks. We also load these weights into a predefined, trusted model architecture (never from user-supplied code).

bernt-matthias replied:
>In the same way you explained here: https://github.com/galaxyproject/tools-iuc/pull/7168#discussion_r2230773776 . Because this is the way Galaxy handles the input files, no?

If biapy requires a specific file extensions for any of the files you need to create a symlink before calling the main program, e.g. `ln -s $mode_selection.selected_mode.cfg_config_process.pretrained_model.biapy_model_path` biapy_model.pth` and then use the symlink.

danifranco replied:
I have just made a new release of BiaPy supporting [safetensors](https://github.com/huggingface/safetensors). In the code here I mader changes to always load and generate .safetensors checkpoints. I'm pushing the changes now.

bernt-matthias reviewed (commented):
Few first comments.

danifranco commented:
Hi,

I still need help with the point 3 and 9. The third one maybe is quick but for the 9 I would need a bit more guidance on how to test the output. 
Thanks a lot!

bgruening commented:
For 3) just assume there is a variable (maybe call if BIAPY_GPU_ENABLED) and its set to "true". If this is there just enable GPU. The Galaxy admin can then set this for this tool if needed.

For 9) I would do the following:

Add a user-input option (type=select multiple=true) and let the user specify which output they want. e.g. charts, binarized images, post-processed images ....

And then you define the output collections like this:

https://github.com/galaxyproject/tools-iuc/blob/bb37578aa61bf4a47af262e02baf0a1c1d9d02c6/tools/hybpiper/hybpiper.xml#L220

Every collection can contain as many files as you want.

I hope that makes sense. Let me know if we can help.

danifranco commented:
I think I have finally corrected everything. What do you think?

danifranco commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,639 @@
+
+        ## Define output directory
+        mkdir -p output &&
+        chmod -R 777 output &&
```
I saw it in some of the tools and thought it was necessary. I'm removing all of them.

danifranco commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,639 @@
+        chmod -R 777 '$checkpoint_dir' &&
+
+        ########## Reuse provided yaml file and update paths ##########
+        #if $mode_selection.selected_mode == 'custom_cfg':
```
I have created 'common_yaml_args' variable to store the common two arguments that were repeated. I think nothing else is repeated...

danifranco commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,639 @@
+            #end if 
+        #end if
+
+        #if $phase in ['train_test', 'train']:
+            ## Error redirection was necessary because chart creation does not always occur at the default interval of every 5 epochs
+            #if 'tcharts' in $outs
+                && mkdir -p train_charts && chmod -R 777 train_charts && 
+                cp output/my_experiment/results/my_experiment_1/charts/*.png train_charts/ 2>/dev/null || true
+            #end if
+            #if 'tlogs' in $outs
+                && mkdir -p train_logs && chmod -R 777 train_logs && 
+                cp output/my_experiment/train_logs/*.txt train_logs/ 2>/dev/null || true
+            #end if
+        #end if
+
+        #if 'checkpoint' in $outs
+            && mkdir -p checkpoints && chmod -R 777 checkpoints && 
+            cp output/my_experiment/checkpoints/*.pth checkpoints/ 2>/dev/null || true
+        #end if
```
For these ones I think we can apply discover_datasets, yes. I'm pushing all the changes soon.

SaimMomin12 commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <edam_topics>
+        <edam_topic>topic_3474</edam_topic>  <!-- Machine learning -->
+        <!-- <edam_topic>topic_3753</edam_topic>  Deep learning (not in EDAM) -->
+    </edam_topics>
+    <edam_operations>
+        <edam_operation>operation_2945</edam_operation>  <!-- Image segmentation -->
+        <edam_operation>operation_3925</edam_operation>  <!-- Object detection -->
+        <edam_operation>operation_3443</edam_operation>  <!-- Image denoising -->
+        <!-- <edam_operation>Single image super-resolution (not in EDAM)</edam_operation> -->
+        <edam_operation>operation_2946</edam_operation>  <!-- Image restoration -->
+        <!-- <edam_operation>Image-to-image translation (not in EDAM)</edam_operation>   -->
+        <!-- <edam_operation>operation_3442</edam_operation> Image classification -->
+        <!-- <edam_operation>Self-supervision learning (not in EDAM)</edam_operation>  -->
+        <edam_operation>operation_2944</edam_operation>  <!-- Image analysis -->
+    </edam_operations>
```
This can go in macros.xml

bernt-matthias replied:
> This can go in macros.xml

Both fine. Up to @danifranco

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,316 @@
+<tool id="biapy" name="Build a workflow with BiaPy" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@" license="MIT">
```
Why should the profile be removed? This is fine IMO.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+<tool id="biapy" name="Build a workflow with BiaPy" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@" license="MIT">
+    <description>: Accessible deep learning on bioimages</description>
```
```suggestion
    <description>Accessible deep learning on bioimages</description>
```

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+                </option>
+            </param>
+            <when value="custom_cfg">
+                <param name="config_path" type="data" format="yaml,txt" optional="false" label="Select a configuration file" help="Input configuration file"/>
```
Are both formats needed?

danifranco replied:
There is no need to of having .yaml or .txt extension. Any plain text will work as long as the content is in yaml format so I put `.yaml` and `.txt` to cover a more range of cases. What should I put?

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+            </param>
+            <when value="custom_cfg">
+                <param name="config_path" type="data" format="yaml,txt" optional="false" label="Select a configuration file" help="Input configuration file"/>
+                <param name="biapy_model_path" type="data" format="data" optional="true" label="Select the model checkpoint (if needed)" help="Path to a pre-trained model checkpoint (.safetensors) generated by BiaPy. Use this only if MODEL.LOAD_CHECKPOINT is set to True in your configuration. Checkpoints are saved in the 'checkpoints' folder inside the results directory."/>
```
https://github.com/galaxyproject/galaxy/pull/20754/ will be merged soon. So we should:

```suggestion
                <param name="biapy_model_path" type="data" format="safetensors" optional="true" label="Select the model checkpoint (if needed)" help="Path to a pre-trained model checkpoint (.safetensors) generated by BiaPy. Use this only if MODEL.LOAD_CHECKPOINT is set to True in your configuration. Checkpoints are saved in the 'checkpoints' folder inside the results directory."/>
```

bernt-matthias replied:
Can the parameter help be more tailored for Galaxy? 

- "path" and "directory" have no meaning for a Galaxy user
-  Can `if MODEL.LOAD_CHECKPOINT is set to True` be reformulated (in terms of the tool form) such that a Galaxy user knows when this is the case.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+                    </param>
+                    <when value="biapy_pretrained">
+                        <!-- Q4 -->
+                        <param name="biapy_model_path" type="data" format="data" optional="false" label="Select the model trained with BiaPy before" help="Specify the path to a pre-trained BiaPy model checkpoint (.safetensors), stored in the 'checkpoints' folder inside the results directory."/>
```
Please already use `safetensors` datatype. 

Analogous to the previous: some terms in the help make little sense to the Galaxy user.

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+                                <option value="torchvision">TorchVision</option>
+                            </param>
+                            <when value="bmz">
+                                <param name="bmz_model_name" type="text" optional="false" value="sensible-cat" label="BioImage Model Zoo model name" help="Enter the name of a pre-trained model from the BioImage Model Zoo (https://bioimage.io/#/models); filter by the BiaPy icon and ensure the model matches your dimensionality (2D/3D) and task (e.g. semantic segmentation).">
```
Just as a note, if https://bioimage.io/#/models would have an API to list model names we could create a select from that... maybe for a next version :)

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+        </collection>
+
+        <collection name="predictions_post_proc" type="list" label="Post-processed test predictions">
+            <discover_datasets directory="post_proc" pattern="(?P&lt;name&gt;.+)\.tif"  format="tif" recurse="false" />
```
If I remember correctly `designation` is preferred over `name` in the pattern.

danifranco replied:
Great, I've jsut changed it!

bernt-matthias commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+    </inputs>
+
+    <outputs>
+        <collection name="predictions_raw" type="list" label="Test predictions">
```
By convention output labels should be prefixed with `${tool.name} on ${on_string}: `. Please add this for all data and collection outputs.

bernt-matthias reviewed (commented):
Hi @danifranco . Thanks for the effort to add safetensors. I gave the tool a last round of reviews and added some (relatively) minor comments. 

Really looking forward to have this in Galaxy in the near future :)

danifranco commented:
I think I have addressed all the comments but the tool is not passing all checks because safetensor datatype is not being merged. So, I assume this will not be merged until that is merged too?

bgruening commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+                && mkdir -p raw && {
+                    ## Instance segmentation
+                    if [ -d "output/my_experiment/results/my_experiment_1/per_image_instances" ]; then
+                        cp output/my_experiment/results/my_experiment_1/per_image_instances/* raw/;
```
is it possible to switch to "mv" here? its on the same harddrive most of the time, so mv should be faster.

bgruening reviewed (commented):
Cool, sorry that it took a while.
We also integrated safetensors here and it now gets merged: 
https://github.com/galaxyproject/galaxy/pull/20754

So I think we are all set. I just had a few last remarks.

Thanks a lot @danifranco

bgruening commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,642 @@
+            --result_dir './output'
+            --name 'my_experiment'
+            --run_id 1
+            #if '$BIAPY_GPU_ENABLED' == 'true':
+                --gpu '0'
+            #end if
```
The #if condition refers to the template language used by Galaxy to create the command line ... so this is evaluated during submission time, NOT runtime - its not a bash "if".

I will make a new suggestion.

bgruening commented on tools/biapy/biapy.xml:
```diff
@@ -0,0 +1,645 @@
+            --result_dir './output'
+            --name 'my_experiment'
+            --run_id 1
+            #if '$BIAPY_GPU_ENABLED' == 'true':
+                --gpu '0'
+            #end if
```
```suggestion
           \${GALAXY_BIAPY_GPU_STRING:-""}
```

What about this. Now the Galaxy admin and Galaxy can create dynamically a bash env var that contains `--gpu 0`

bgruening replied:
If this env does not exists the string is empty. So no --gpu param et all.

danifranco replied:
It was crashing so I've just the upload this line with: `--gpu \${GALAXY_BIAPY_GPU_STRING:-""}` so it is empty when nothing is set. Whenever you want to use a GPU you can then set `GALAXY_BIAPY_GPU_STRING` with an integer there (or multiple ints separated by comma e.g. `"0,1,2"`)

bgruening reviewed (approved):
Cool! Very nice first version! Lets get this deployed :)

If you feel like it: https://galaxyproject.eu/posts/2020/08/22/three-steps-to-galaxify-your-tool/

Step 3 :)
